src/execution/compiler/operator/hash_aggregation_translator.cpp

```diff
@@ -426,28 +436,47 @@ void HashAggregationTranslator::PerformPipelineWork(WorkContext *context, Functi
       // function parameter; specifically, the last argument in the worker
       // function which we're generating right now. Pull it out.
       auto agg_ht_param_position = GetPipeline()->PipelineParams().size();
-      auto agg_ht = function->GetParameterByPosition(agg_ht_param_position);
+      agg_ht = function->GetParameterByPosition(agg_ht_param_position);
       ScanAggregationHashTable(context, function, agg_ht);
     } else {
-      ScanAggregationHashTable(context, function, global_agg_ht_.GetPtr(codegen));
+      agg_ht = global_agg_ht_.GetPtr(codegen);
+      ScanAggregationHashTable(context, function, agg_ht);
     }
   }
 }
 
 void HashAggregationTranslator::FinishPipelineWork(const Pipeline &pipeline, FunctionBuilder *function) const {
-  if (IsBuildPipeline(pipeline) && build_pipeline_.IsParallel()) {
-    auto *codegen = GetCodeGen();
-    auto global_agg_ht = global_agg_ht_.GetPtr(codegen);
-    auto thread_state_container = GetThreadStateContainer();
-    auto tl_agg_ht_offset = local_agg_ht_.OffsetFromState(codegen);
-    function->Append(codegen->AggHashTableMovePartitions(global_agg_ht, thread_state_container, tl_agg_ht_offset,
-                                                         merge_partitions_fn_));
+  auto *codegen = GetCodeGen();
+  const auto &agg_ht = build_pipeline_.IsParallel() ? local_agg_ht_ : global_agg_ht_;
+
+  if (IsBuildPipeline(pipeline)) {
+    if (build_pipeline_.IsParallel()) {
+      auto global_agg_ht = global_agg_ht_.GetPtr(codegen);
+      auto thread_state_container = GetThreadStateContainer();
+      auto tl_agg_ht_offset = local_agg_ht_.OffsetFromState(codegen);
+      function->Append(codegen->AggHashTableMovePartitions(global_agg_ht, thread_state_container, tl_agg_ht_offset,
+                                                           merge_partitions_fn_));
+    }
+
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_BUILD,
+                  brain::ExecutionOperatingUnitFeatureAttribute::NUM_ROWS, pipeline, CounterVal(num_agg_inputs_));
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_BUILD,
+                  brain::ExecutionOperatingUnitFeatureAttribute::CARDINALITY, pipeline,
+                  codegen->CallBuiltin(ast::Builtin::AggHashTableGetTupleCount, {agg_ht.GetPtr(codegen)}));
+    FeatureArithmeticRecordMul(function, pipeline, GetTranslatorId(), CounterVal(num_agg_inputs_));
+  } else {
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_ITERATE,
+                  brain::ExecutionOperatingUnitFeatureAttribute::NUM_ROWS, pipeline, CounterVal(num_agg_outputs_));
+    FeatureRecord(function, brain::ExecutionOperatingUnitType::AGGREGATE_ITERATE,
+                  brain::ExecutionOperatingUnitFeatureAttribute::CARDINALITY, pipeline,
+                  codegen->CallBuiltin(ast::Builtin::AggHashTableGetTupleCount, {agg_ht.GetPtr(codegen)}));
```

Comment:
Let's add the multiplication back here...

Reply:
Added it to FinishPipelineWork (the convention I'm running with now is for all recording to happen there by default).

Your reasoning makes sense, the translator ID is the same but the pipeline ID is not the same.